File: app/src/main/java/io/plan8/backoffice/adapter/RestfulAdapter.java
package io.plan8.backoffice.adapter;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import io.plan8.backoffice.ApplicationManager;
import io.plan8.backoffice.api.ApiService;
import io.plan8.backoffice.util.AutoFillHeaderInterceptor;
import io.plan8.backoffice.util.ConnectivityInterceptor;
import io.plan8.backoffice.util.NullOnEmptyConverterFactory;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by SSozi on 2017. 11. 28..
 */

public class RestfulAdapter {
    private static volatile RestfulAdapter instance = null;
    private ApiService needTokenApiService;
    private ApiService apiService;
    private Context context;

    public RestfulAdapter() {
    }

    public RestfulAdapter(Context context) {
        this.context = context;
    }

    public static RestfulAdapter getInstance() {
        if (null == instance) {
            synchronized (RestfulAdapter.class) {
                instance = new RestfulAdapter();
            }
        }
        return instance;
    }

    public ApiService getNeedTokenApiService() {
        if (needTokenApiService == null) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(interceptor)
                    .addInterceptor(new AutoFillHeaderInterceptor(context)).build();

            Gson gson = new GsonBuilder()
                    .setLenient()
                    .excludeFieldsWithoutExposeAnnotation()
                    .create();

            /**
             *  api - sandbox
             */
//                //Retrofit 설정
//                needTokenApiService = new Retrofit.Builder()
//                        .baseUrl(Constants.API_SANDBOX_URL)
//                        .client(client)
//                        .addConverterFactory(GsonConverterFactory.create(gson))
//                        .build().create(ApiService.class); //인터페이스 연결

            /******
             *  api - production
             */
            //Retrofit 설정
            needTokenApiService = new Retrofit.Builder()
                    .baseUrl(ApplicationManager.getInstance().getServerUrl())
                    .client(client)
                    .addConverterFactory(new NullOnEmptyConverterFactory())
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build().create(ApiService.class); //인터페이스 연결

            return needTokenApiService;
        } else {
            return needTokenApiService;
        }
    }

    public ApiService getApiService() {
        if (apiService == null) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(interceptor)
                    .addInterceptor(new ConnectivityInterceptor(context)).build();

            Gson gson = new GsonBuilder()
                    .setLenient()
                    .excludeFieldsWithoutExposeAnnotation()
                    .create();

            /**
             *  api - sandbox
             */
//                //Retrofit 설정
//                needTokenApiService = new Retrofit.Builder()
//                        .baseUrl(Constants.API_SANDBOX_URL)
//                        .client(client)
//                        .addConverterFactory(GsonConverterFactory.create(gson))
//                        .build().create(ApiService.class); //인터페이스 연결

            /******
             *  api - production
             */
            //Retrofit 설정
            apiService = new Retrofit.Builder()
                    .baseUrl(ApplicationManager.getInstance().getServerUrl())
                    .client(client)
                    .addConverterFactory(new NullOnEmptyConverterFactory())
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build().create(ApiService.class); //인터페이스 연결

            return apiService;
        } else {
            return apiService;
        }
    }

    public Context getContext() {
        return context;
    }

    public void setContext(Context context) {
        this.context = context;
    }
}

File: app/src/main/java/io/plan8/backoffice/adapter/RestfulAdapter.kt
package io.plan8.backoffice.adapter

import android.annotation.SuppressLint
import android.content.Context

import com.google.gson.GsonBuilder
import io.plan8.backoffice.ApplicationManager

import io.plan8.backoffice.api.ApiService
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * Created by SSozi on 2017. 11. 7..
 */

class RestfulAdapter {
    private lateinit var context: Context
    private lateinit var token: String

    companion object {
        @SuppressLint("StaticFieldLeak")
        var instance: RestfulAdapter? = null
        var retrofitServiceApi: ApiService? = null

        @Synchronized
        fun build(context: Context): RestfulAdapter {
            if (instance == null) instance = RestfulAdapter(context)
            return instance as RestfulAdapter
        }
    }

    constructor()

    constructor(context: Context) {
        this.context = context
//        this.token = token
    }

    //Retrofit 설정
    //인터페이스 연결
    val serviceApi: ApiService?
        get() {
            if (retrofitServiceApi == null) {
                val client = OkHttpClient.Builder().build()

                val gson = GsonBuilder()
                        .setLenient()
                        .create()
                retrofitServiceApi = Retrofit.Builder()
                        .baseUrl(ApplicationManager.BASE_URL)
                        .client(client)
                        .addConverterFactory(GsonConverterFactory.create(gson))
                        .build().create(ApiService::class.java)
            }
            return retrofitServiceApi
        }

    val authServiceApi: ApiService?
        get() {
            if (retrofitServiceApi == null) {
                val client: OkHttpClient.Builder? = OkHttpClient().newBuilder().addInterceptor(object : Interceptor{
                    override fun intercept(chain: Interceptor.Chain?): Response {
                        val request: Request = chain!!.request().newBuilder().addHeader("authorization", token).build()
                        return chain.proceed(request)
                    }

                })


                val gson = GsonBuilder()
                        .setLenient()
                        .create()
//                retrofitServiceApi = Retrofit.Builder()
//                        .baseUrl(ApplicationManager.BASE_URL)
//                        .client(client!!)
//                        .addConverterFactory(GsonConverterFactory.create(gson))
//                        .build().create(ApiService::class.java)
            }
            return retrofitServiceApi
        }
}
